src/Language/PureScript/TypeChecker.hs

```diff
@@ -155,18 +155,28 @@ checkDuplicateTypeArguments args = for_ firstDup $ \dup ->
   firstDup = listToMaybe $ args \\ nub args
 
 checkTypeClassInstance
-  :: (MonadState CheckState m, MonadError MultipleErrors m, MonadWriter MultipleErrors m)
-  => ModuleName
+  :: (MonadState CheckState m, MonadError MultipleErrors m)
+  => TypeClassData
+  -> Int -- type argument index
   -> Type
   -> m ()
-checkTypeClassInstance _ (TypeVar _) = return ()
-checkTypeClassInstance _ (TypeLevelString _) = return ()
-checkTypeClassInstance _ (TypeConstructor ctor) = do
-  env <- getEnv
-  when (ctor `M.member` typeSynonyms env) . throwError . errorMessage $ TypeSynonymInstance
-  return ()
-checkTypeClassInstance m (TypeApp t1 t2) = checkTypeClassInstance m t1 >> checkTypeClassInstance m t2
-checkTypeClassInstance _ ty = throwError . errorMessage $ InvalidInstanceHead ty
+checkTypeClassInstance cls i = check where
+  -- is this argument fully determined via fundeps
+  isFunDepDetermined = (Any False, Any True) == foldMap determining (typeClassDependencies cls)
+    where determining fd = (Any (i `elem` fdDeterminers fd),
```

Comment:
I'm not sure that makes sense 😕, wouldn't that mean `x -> y, y -> z` and `x -> y z` are different?  Using that algorithm:

Walkthrough of checking `z` in `x -> y z`
```
determined by: x ("determiners for that fundep")
x isn't determined by anything
therefore, c is determined
```

Walkthrough of checking `z` in `x -> y, y -> z`:
```
determined by: y ("determiners for that fundep", from: y -> z)
y is determined by: x ("determined by some other fundep")
therefore, z is not determined ???
```

I have come up with what I think is a correct general solution: I'll push a new commit in a minute for you to look at.  It uses a graph to compute contributing dependencies between variables.  We will want to move computing the determined args out of the instance check, but it's there for now whilst we discuss the algorithm.